src/balance.rs

```diff
@@ -26,41 +41,219 @@ impl<T: Config> From<Error<T>> for StatusCode {
+
+        {
+            let mut balances = self.balances.borrow_mut();
+            balances.insert(account.clone(), *balance);
+        }
+        {
+            let mut initial_state = self.initial_state.borrow_mut();
+            initial_state.insert(account.clone(), *balance);
+        }
+
+        Ok(())
+    }
+
+    /// Executes the true transactions on the blockchain/substrate side after execution of
+    /// Move-script.
+    pub fn apply_transactions(&self) -> DispatchResult {
+        use core::cmp::Ordering;
+
+        self.cmp_with_initial_state()?;
+
+        let balances = self.balances.borrow();
+        let mut purse = NegativeImbalanceOf::<T>::zero();
+        let mut depts = Vec::<(&AccountIdOf<T>, BalanceOf<T>)>::new();
+
+        // Calculate balance differences and withdraw negative ones from user's accounts.
+        for (account, balance) in balances.iter() {
+            let initial_state = self.initial_state.borrow();
+            let true_balance = initial_state.get(account).unwrap();
+            match (*true_balance).cmp(balance) {
+                Ordering::Greater => {
+                    let dept = *true_balance - *balance;
+                    let imbalance = T::Currency::withdraw(
+                        account,
+                        dept,
+                        WithdrawReasons::TRANSFER,
+                        ExistenceRequirement::AllowDeath,
+                    )?;
+                    purse = purse.merge(imbalance);
+                }
+                Ordering::Less => {
+                    let dept = *balance - *true_balance;
+                    depts.push((account, dept));
+                }
+                Ordering::Equal => {}
+            }
+        }
+
+        // Now deposit depts from purse to new owners.
+        for (account, balance) in depts.into_iter() {
+            let imbalance = purse.extract(balance);
+            T::Currency::resolve_creating(account, imbalance);
+        }
+
+        Ok(())
+    }
+
+    /// Ensures that user can withdraw that given amount of money, which eventually will be used
+    /// within the Move-script execution.
+    #[allow(dead_code)]
+    fn ensure_can_withdraw(
+        &self,
+        account: &AccountIdOf<T>,
+        amount: &BalanceOf<T>,
+    ) -> DispatchResult {
+        if *amount < T::Currency::free_balance(account) {
+            Ok(())
+        } else {
+            Err(Error::<T>::InsufficientBalance.into())
+        }
+    }
+
+    /// Method ensures that a certain user exists in our balances record.
+    fn ensure_account_exists(&self, user: &AccountIdOf<T>) {
+        if !self.balances.borrow().contains_key(user) {
+            let balance = BalanceOf::<T>::from(0u128);
+            let _ = self.balances.borrow_mut().insert(user.clone(), balance);
+            let _ = self
+                .initial_state
+                .borrow_mut()
+                .insert(user.clone(), balance);
```

Comment:
not really needed, no?

Reply:
Please check again, not sure, if this is still an issue, or got changed with `HashMap<AccountId, Balance>`... :thinking: :blush: